src/inventory.rs

```diff
@@ -0,0 +1,303 @@
+
+    fn mark_dirty(&mut self, dirty: bool) {
+        self.dirty = dirty
+    }
+
+    fn is_dirty(&self) -> bool {
+        self.dirty
+    }
+}
+
+/// Represents what the player sees when they open an object's Inventory.
+///
+/// This exists because when an object inventory screen is being shown to the
+/// player, it also shows part of the player's inventory so they can move items
+/// between the inventories.
+pub struct WindowInventory {
+    pub window_id: u8,
+    pub object_inventory: InventoryId,
+}
+
+impl WindowInventory {
+    pub fn new(window_id: impl Into<u8>, object_inventory: InventoryId) -> Self {
+        WindowInventory {
+            window_id: window_id.into(),
+            object_inventory,
+        }
+    }
+
+    fn slots(
+        &self,
+        obj_inventory: &ConfigurableInventory,
+        player_inventory: &PlayerInventory,
+    ) -> Vec<Slot> {
+        let total_slots = obj_inventory.slots.len() + PlayerInventory::GENERAL_SLOTS.len();
+        (0..total_slots)
+            .map(|s| {
+                if s < obj_inventory.slot_count() {
+                    return obj_inventory.slot(s as SlotId);
+                }
+                let offset = obj_inventory.slot_count();
+                player_inventory.slot((s - offset) as SlotId + PlayerInventory::GENERAL_SLOTS.start)
+            })
+            .collect()
+    }
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Default, Debug)]
+pub struct InventoryId(Key);
+
+/// Manages all inventories that are present in the server.
+pub struct Inventories {
+    slab: VersionedSlab<ConfigurableInventory>,
+}
+
+impl Default for Inventories {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Inventories {
+    pub fn new() -> Self {
+        Self {
+            slab: VersionedSlab::new(),
+        }
+    }
+
+    /// Creates a new inventory on a server.
+    pub fn insert(
+        &mut self,
+        inv: ConfigurableInventory,
+    ) -> (InventoryId, &mut ConfigurableInventory) {
+        let (key, value) = self.slab.insert(inv);
+        (InventoryId(key), value)
+    }
+
+    /// Removes an inventory from the server.
+    pub fn remove(&mut self, inv: InventoryId) -> Option<ConfigurableInventory> {
+        self.slab.remove(inv.0)
+    }
+
+    /// Returns the number of inventories in this container.
+    pub fn len(&self) -> usize {
+        self.slab.len()
+    }
+
+    /// Returns `true` if there are no inventories.
+    pub fn is_empty(&self) -> bool {
+        self.slab.len() == 0
+    }
+
+    pub fn get(&self, inv: InventoryId) -> Option<&ConfigurableInventory> {
+        self.slab.get(inv.0)
+    }
+
+    pub fn get_mut(&mut self, inv: InventoryId) -> Option<&mut ConfigurableInventory> {
+        self.slab.get_mut(inv.0)
+    }
+
+    pub(crate) fn sync<C: Config>(&mut self, clients: &mut Clients<C>) {
+        // sync open, dirty inventories to clients
+        let _obj_inventories_cleaned: HashSet<InventoryId> = clients
```

Comment:
We need to do this every tick to keep inventories in sync with clients. This is very much not the most optimal way to do it, but it is the most simple. Using [SetContainerSlot](https://wiki.vg/Protocol#Set_Container_Slot) instead of shoving the whole inventory down the pipe is on my todo list for later.

As for collecting, we need the collect because iterators are lazy by default. The HashSet part is left over from a previous implementation, I'll fix that.